ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/metrics/ConsumptionSpeedMetric.scala

```diff
@@ -4,64 +4,117 @@
 
 import java.time.Instant
 
 final case class ConsumptionSpeedMetric[T](
     periodMillis: Long,
     recordTimeFunction: T => Seq[Timestamp],
-    firstRecordTime: Option[Instant] = None,
-    lastRecordTime: Option[Instant] = None,
+    objectives: Map[ServiceLevelObjective[ConsumptionSpeedMetric.Value], Option[
+      ConsumptionSpeedMetric.Value
+    ]],
+    previousLatest: Option[Instant] = None,
+    currentPeriodLatest: Option[Instant] = None,
 ) extends Metric[T] {
+  import ConsumptionSpeedMetric._
 
-  override type Value = ConsumptionSpeedMetric.Value
+  override type V = Value
+  override type Objective = ServiceLevelObjective[Value]
 
   override def onNext(value: T): ConsumptionSpeedMetric[T] = {
     val recordTimes = recordTimeFunction(value)
-    val updatedFirstRecordTime =
-      firstRecordTime match {
+    val newPreviousLatest =
+      previousLatest match {
         case None =>
           recordTimes.headOption.map { recordTime =>
             Instant.ofEpochSecond(recordTime.seconds.toLong, recordTime.nanos.toLong)
           }
-        case recordTime => recordTime
+        case v => v
       }
-    val updatedLastRecordTime = recordTimes.lastOption.map { recordTime =>
+    val newCurrentPeriodLatest = recordTimes.lastOption.map { recordTime =>
       Instant.ofEpochSecond(recordTime.seconds.toLong, recordTime.nanos.toLong)
     }
+
     this.copy(
-      firstRecordTime = updatedFirstRecordTime,
-      lastRecordTime = updatedLastRecordTime,
+      previousLatest = newPreviousLatest,
+      currentPeriodLatest = newCurrentPeriodLatest,
     )
   }
 
-  override def periodicValue(): (Metric[T], ConsumptionSpeedMetric.Value) = {
-    val updatedMetric = this.copy(firstRecordTime = None, lastRecordTime = None)
-    (updatedMetric, ConsumptionSpeedMetric.Value(periodicSpeed))
+  override def periodicValue(): (Metric[T], Value) = {
+    val value = Value(periodicSpeed)
+    val updatedMetric = this.copy(
+      previousLatest = if (currentPeriodLatest.isDefined) currentPeriodLatest else previousLatest,
+      currentPeriodLatest = None,
+      objectives = updatedObjectives(value),
+    )
+    (updatedMetric, value)
   }
 
-  override def finalValue(totalDurationSeconds: Double): ConsumptionSpeedMetric.Value =
-    ConsumptionSpeedMetric.Value(None)
+  override def finalValue(totalDurationSeconds: Double): Value =
+    Value(None)
+
+  override def violatedObjectives: Map[ServiceLevelObjective[Value], Value] =
+    objectives
+      .collect {
+        case (objective, value) if value.isDefined => objective -> value.get
+      }
 
   private def periodicSpeed: Option[Double] =
-    (firstRecordTime, lastRecordTime) match {
-      case (Some(first), Some(last)) =>
-        Some((last.toEpochMilli - first.toEpochMilli) * 1.0 / periodMillis)
+    (previousLatest, currentPeriodLatest) match {
+      case (Some(previous), Some(current)) =>
+        Some((current.toEpochMilli - previous.toEpochMilli).toDouble / periodMillis)
```

Comment:
There is no guarantee that the periodMillis is being maintained by the thread scheduler. It is safer to assume it is not and rely on a current time, to facilitate the unit testing you may want to supply current time as argument of periodicValue(periodDuration: Duration)

Reply:
Good catch! I'll fix this metric and others that may rely on the constant period duration.